### DISCUSSION TITLE
Finding epsilon

### DISCUSSION THREAD

**Discussion by sidify** [opendp/opendp#436] [SmartNoise] [answered]
Suppose I create a synthetic dataset out of a real dataset using some epsilon. How can I verify that my synthetic dataset is atleast epsilon differential private?

**Comment by Shoeboxam** [comment]
Are you trying to work out how much epsilon was used by comparing a synthetic dataset to the real dataset? I don't think there's a robust way to validate this. 

If you are generating your synthetic dataset via histograms, it would be easier to run a stochastic evaluation to compare the empirical distribution of histogram releases on `X` and any `X'` chosen at random. On average, answers should differ by no more than `e^epsilon`. The synthetic part follows by postprocessing.

**Comment by sidify** [comment]
Hi @Shoeboxam , thank you for your response. Yes, I am trying to validate that in the end my synthetic data is atleast `Ɛ` differential private. For example, it maybe a responsibility of a data quality expert (independent of the data curator) within the organisation to verify that the synthetic data is up to the mark. The data quality expert has only access to both the real as well as the synthetic dataset. 

I am trying to understand, this would mean that if we take an example query `Q` when applied to dataset `D` and `D'` (differ only in one record), should give me a result within the differential private bounds, basically `P[ Q(D) = S ] ≤ e^Ɛ * P[ Q(D') = S ]`  should hold true.

But then, there are too many possibilities of `D' ` and also infinite queries (`Q`s) possible. Given this scenario, how can we ascertain a data quality expert?

**Comment by joshua-oss** [comment, marked as answer]
The paper at [1] outlines an approach for estimating epsilon when the output of the randomized algorithm is high-dimensional and computationally expensive to produce.  The main contributions are in providing tighter bounds for epsilon with fewer runs of the algorithm needed.  That paper is focused on machine learning models, but the technique is easier to apply in the setting of synthetic data.   Synthetic data represents a joint distribution, which allows you to skip all of the shadow models and directly perform the likelihood test from section 4.2 of [2].  We have applied this approach with synthetic data, and it seems to work very well.  We plan to release code that performs these tests against synthetic data and private synopses, as part of `smartnoise-synth`, in the coming months.

Prior to the work above, we also used the "stochastic tester" described on page 18 of [3] for basic testing on synthetic data.  The way to apply that approach in the synthetic data setting is to run the synthesizer multiple times, and then feed the output of some query over the data (e.g. marginal counts, or means) into the stochastic testing algorithm.  That's not ideal, since the approach is designed for single summary statistics, but it can find serious privacy bugs in the synthesizer.

[1] https://arxiv.org/pdf/2206.05199.pdf
[2] https://arxiv.org/pdf/2111.09679.pdf
[3] https://arxiv.org/pdf/1909.01917.pdf
